drivers/clock_control/clock_stm32_mco.c

```diff
@@ -0,0 +1,136 @@
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_SYSCLK
+#else
+#error "MCO1 source not a valid clock source"
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(mco1), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(mco2), okay)
+#if DT_PROP(DT_NODELABEL(mco2), source) == NO_SEL
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_SYSCLK
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLLI2S
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLLI2S
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_HSE
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_HSE
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_LSI
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_LSI
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_CSI
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_CSI
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLLCLK
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLLCLK
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLLPCLK
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLL1PCLK
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLL2PCLK
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLL2PCLK
+#else
+#error "MCO2 source not a valid clock source"
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(mco2), okay) */
+
+#define fn_mco_prescaler(i, v) LL_RCC_MCO ## i ## _DIV_ ## v
+#define mco_prescaler(i, v) fn_mco_prescaler(i, v)
+
+#define fn_mco_source(i) MCO ## i ## _SOURCE
+#define mco_source(i) fn_mco_source(i)
+
+struct stm32_mco_config {
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct stm32_mco_data {
+	uint32_t source;
+	uint32_t prescaler;
+};
+
+static int stm32_mco_init(const struct device *dev)
+{
+	const struct stm32_mco_config *config = dev->config;
+	struct stm32_mco_data *data = dev->data;
+
+	(void)pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	LL_RCC_ConfigMCO(data->source);
+#else
+	LL_RCC_ConfigMCO(data->source, data->prescaler);
+#endif
+
+	return 0;
+}
+
+#define STM32_MCO_INIT(inst)                                            \
+									\
+PINCTRL_DT_INST_DEFINE(inst);                                           \
+									\
+const static struct stm32_mco_config stm32_mco_config_##inst = {        \
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),                   \
+};                                                                      \
+static struct stm32_mco_data stm32_mco_data_##inst = {                  \
+	.source = mco_source(DT_INST_REG_ADDR(inst)),                   \
+	.prescaler = mco_prescaler(DT_INST_REG_ADDR(inst),              \
+				   DT_INST_PROP(inst, prescaler)),      \
+};                                                                      \
+									\
+DEVICE_DT_INST_DEFINE(inst, stm32_mco_init, NULL,                       \
+	&stm32_mco_data_##inst,                                         \
+	&stm32_mco_config_##inst,                                       \
+	PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,               \
```

Comment:
This follows existing behavior so this is ok, but ideally this should happen before clock control driver initialization.
See https://github.com/zephyrproject-rtos/zephyr/issues/31912#issuecomment-1334061384

Reply:
Do you mind explaining why it needs to be done in this order. The init priority is chosen so that the pinctrl apply function will work.
Because doing that before GPIO is initialized will not work.

So then we need two different priority levels.

Reply:
> The init priority is chosen so that the pinctrl apply function will work.

Good point.

> Do you mind explaining why it needs to be done in this order ?

According to RM, this is all about avoiding glitches.
`Clock source selection may generate glitches on MCO2.`
`Modification of this prescaler may generate glitches on MCO2`

So it depends if we say that avoiding glitches worthes the cost of having two steps of initialization for this driver. As I said, the proposed implementation is close to existing behavior in that matter, so it can remain as is. 
If required, fixing this behavior could be done later (as transparent for user).

Reply:
Alright. Agree to fix it later as it is replaces the current behavior. Good to keep the changes in this PR down.